ion-schema/src/nfa.rs

```diff
@@ -0,0 +1,258 @@
+use crate::ion_path::IonPath;
+use crate::system::{TypeId, TypeStore};
+use crate::types::TypeValidator;
+use crate::IonSchemaElement;
+use ion_rs::value::owned::Element;
+use std::collections::{HashMap, HashSet};
+use std::rc::Rc;
+
+/// Represents an event in NFA which can either be an Ion Value or an end of stream
+#[derive(Debug, Clone, PartialEq)]
+pub enum Event {
+    Value(Element),
+    EndOfStream,
+}
+
+impl Event {
+    // Note: can not use this as a From<Element> implementation as Element is not defined within this crate
+    pub(crate) fn from_elements_to_events<T: IntoIterator<Item = Element>>(iter: T) -> Vec<Event> {
+        let mut events: Vec<Event> = iter
+            .into_iter()
+            .map(<Element as Into<Event>>::into)
+            .collect();
+        // add END_OF_STREAM at the end of events vec
+        events.push(Event::EndOfStream);
+        events
+    }
+}
+
+impl From<Element> for Event {
+    fn from(value: Element) -> Self {
+        Event::Value(value)
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct Transition {
+    destination: usize,      // represents destination state
+    to_optional_state: bool, // represents if destination state is an optional state or not
+    type_id: Option<TypeId>, // None is used for final state transition
+}
+
+/// Represents an NFA that will eb used by the ordered_elements` constraint in order to validate an Ion Value
+#[derive(Default, Debug, Clone, PartialEq)]
+pub struct Nfa {
+    pub(crate) states: Vec<State>, // represents states with state_id
+    pub(crate) transitions: HashMap<usize, HashSet<Transition>>, // represents transitions between states
+}
+
+/// This is a context which will be used while validating an Ion value for `ordered_elements` constraint using its NFA
+#[derive(Debug, Clone)]
+pub struct NfaRun {
+    pub(crate) visits: HashMap<usize, usize>, // represents a map of (stat_id, visit_count)
+    pub(crate) nfa: Rc<Nfa>,                  // reference to the NFA
```

Comment:
nitpick—this comment doesn't actually say anything useful (since we already know from the type name that `Rc<Nfa>` is a reference to the NFA)